tools/upscaling_coefficients/upscaler_demo.py

```diff
@@ -0,0 +1,793 @@
+    -0.008138951872408831, -0.0015349087147535293], [-0.019308217274971017,
+    -0.03620398561701561, -0.019741250657301434, -0.03919545281633189,
+    -0.014560933634183596], [-0.03866988475938143, 0.1422954970729258,
+    0.5659339775075574, 0.0804518075119682, -0.028882977402423963],
+    [-0.035832545662066145, 0.11572472115297625, 0.47416733354946283,
+    0.06284440084948135, -0.026850659249274097], [-0.01626392539751837,
+    -0.039541478550530786, -0.04046620032608076, -0.03979621423581152,
+    -0.012244853215160443]], [[-0.0015349087147535293, -0.008138951872408831,
+    -0.013176682690936201, -0.008907053401106528, -0.00210156216711573],
+    [-0.0145609336341836, -0.0391954528163319, -0.019741250657301437,
+    -0.03620398561701563, -0.01930821727497102], [-0.028882977402423963,
+    0.08045180751196825, 0.5659339775075575, 0.14229549707292577,
+    -0.038669884759381434], [-0.0268506592492741, 0.06284440084948138,
+    0.47416733354946305, 0.11572472115297622, -0.03583254566206615],
+    [-0.012244853215160445, -0.03979621423581153, -0.04046620032608077,
+    -0.03954147855053079, -0.016263925397518374]], [[-0.0010778269600400273,
+    -0.00730937211152405, -0.01288326617180421, -0.00957114638494601,
+    -0.002791220988040242], [-0.010740155274818482, -0.04044013741654317,
+    -0.020308275361446707, -0.031939432194582666, -0.024336140047717],
+    [-0.020702111700092594, 0.03369473939392682, 0.5398093391410693,
+    0.21238920010551754, -0.04562658821385713], [-0.019367833372750342,
+    0.022877571178541396, 0.452206427023829, 0.17567618137781177,
+    -0.04246845133905895], [-0.009019734953997752, -0.039000354140279844,
+    -0.040021010867420236, -0.03838632575427138, -0.02066407496750483]],
+    [[-0.0007159165805851705, -0.006389875713059274, -0.012319067611648212,
+    -0.010202069931803576, -0.0036010963945262565], [-0.007755474877032563,
+    -0.03942402104403912, -0.02118180471068666, -0.0278857406191104,
+    -0.029204772772557758], [-0.015459264122748742, -7.890574314417695e-05,
+    0.4909386897413151, 0.2868949551896562, -0.04562755195174026],
+    [-0.01450239495172347, -0.0057304550220541465, 0.4111930051936302,
+    0.23903219477825297, -0.04314593981787025], [-0.0064664894229144,
+    -0.03663285147762567, -0.03901507994141055, -0.037526010691823306,
+    -0.025194061504750526]], [[-0.0004507227386051219, -0.005422291349995998,
+    -0.01155272937910007, -0.01081748428801323, -0.00447631814814665],
+    [-0.005422291349995998, -0.03680917952171095, -0.022489385288542607,
+    -0.024475522375569655, -0.03324558280295302], [-0.011552729379100074,
+    -0.02248938528854261, 0.4272005024152728, 0.3575770812164143,
+    -0.0378381168526885], [-0.010817484288013228, -0.02447552237556967,
+    0.3575770812164143, 0.29895328454745285, -0.03706352644207268],
+    [-0.00447631814814665, -0.03324558280295302, -0.0378381168526885,
+    -0.03706352644207269, -0.029286133281073243]]]]
+}
```

Comment:
OK for a demo, but overall, there might be better alternatives to cluttering the source file with a 700-lines constant. Would it work (for example) to use a .npz instead?

>>> numpy.savez_compressed('/tmp/data.npz', k2=numpy.zeros([4, 4]), k4=numpy.zeros([4, 4]), k8=numpy.zeros([4, 4]))
>>> d = numpy.load('/tmp/data.npz')
>>> type(d)
<class 'numpy.lib.npyio.NpzFile'>
>>> d
<numpy.lib.npyio.NpzFile object at 0x7f0328f523a0>
>>> list(d)
['k2', 'k4', 'k8']
>>> d['k2'].shape
(4, 4)
>>> d['k2'][:2, :2]
array([[0., 0.],
       [0., 0.]])

Reply:
Another alternative would be to move these to a separate upscaler_demo_constants.py.

A third alternative - and this is actually sort-of violating the letter of some readability guidelines here, but in this case helps quite a bit with readability, is to use this technique:

===
#!/usr/bin/env python3

FOO_CONSTANTS = _get_constants()

... code ...

def main():
  ...

if __name__ == '__main__':
  main()

def _get_constants():
  # all the distracting stuff goes here, at the very end
===

This sort-of emulates what Perl's __DATA__ achieves, see e.g.: https://perldoc.perl.org/SelfLoader#The-__DATA__-token

I would say there is some value in not having this big wall of numbers at the start of the file.

Reply:
Makes sense, moved the constants to the bottom, in a way similar to what you suggested.